webapp-django/crashstats/authentication/views.py

```diff
@@ -49,3 +69,85 @@ def debug_login(request):
     future = datetime.datetime.utcnow() + datetime.timedelta(seconds=10)
     response.set_cookie('cookie_value', cookie_value, expires=future)
     return response
+
+
+@json_view
+def oauth2_signout(request):
+    """If people click the "Sign out" link, an AJAX post will trigger
+    a post to this view. But if the user right-clicks to, for example,
+    "Open Link in New Tab" they'll arrive here.
+
+    If they arrive here, in any way, and are NOT signed in, redirect
+    out.
+    """
+    if not request.user.is_authenticated():
+        return redirect('/')
+
+    if request.method == 'POST':
+        auth.logout(request)
+        return {'OK': True}
+
+    return render(request, 'auth/signout.html')
+
+
+@require_POST
+@json_view
+def oauth2_signin(request):
+    token = request.POST['token']
+    try:
+        idinfo = client.verify_id_token(
+            token,
+            settings.OAUTH2_CLIENT_ID
+        )
+    except crypt.AppIdentityError:
+        return http.HttpResponseForbidden(
+            'Invalid Identity token'
+        )
+
+    # We should probably look at idinfo['email_verified'] and
+    # return a Forbidden if it's not verified.
+    if idinfo['aud'] != settings.OAUTH2_CLIENT_ID:
+        return http.HttpResponseForbidden(
+            'Invalid Client ID ({})'.format(idinfo['aud'])
+        )
+    if idinfo['iss'] not in settings.OAUTH2_VALID_ISSUERS:
+        return http.HttpResponseForbidden(
+            'Invalid issuer ({})'.format(idinfo['iss'])
+        )
+    # log in the user
+    try:
+        user = User.objects.get(email__iexact=idinfo['email'])
+        if not user.is_active:
+            return http.HttpResponseForbidden(
+                'Inactivated user'
+            )
+        # We already had a user by that email address, but if the
+        # first or last name is different, override what we already had.
+        if (
+            idinfo.get('given_name') is not None and
+            user.first_name != idinfo['given_name']
+        ):
+            user.first_name = idinfo['given_name']
+            user.save()
+        if (
+            idinfo.get('family_name') is not None and
+            user.last_name != idinfo['family_name']
+        ):
+            user.last_name = idinfo['family_name']
+            user.save()
+    except User.DoesNotExist:
+        user = User.objects.create(
+            username=default_username(idinfo['email']),
+            email=idinfo['email'],
+            first_name=idinfo.get('given_name', ''),
+            last_name=idinfo.get('family_name', ''),
+        )
+    user.backend = 'django.contrib.auth.backends.ModelBackend'
+    request.user = user
```

Comment:
Can you explain these last two lines? In particular, I don't understand why you want to set the user in ``request`` before calling ``auth.login(request, user)``.